programs/editor/bellabello.js

```diff
@@ -0,0 +1,498 @@
+                    var canvasContainerElement;
+
+                    editor = e;
+
+                    canvasContainerElement = editor.getCanvasContainerElement();
+                    savingOverlay = addStatusOverlay(canvasContainerElement, "document-save.png", 0);
+                    hasLocalUnsyncedOpsOverlay = addStatusOverlay(canvasContainerElement, "vcs-locally-modified.png", 0);
+                    disconnectedOverlay = addStatusOverlay(canvasContainerElement, "network-disconnect.png", 1);
+
+                    editor.addEventListener(Wodo.EVENT_BEFORESAVETOFILE, function() {
+                        savingOverlay.style.display = "";
+                    });
+                    editor.addEventListener(Wodo.EVENT_SAVEDTOFILE, function() {
+                        savingOverlay.style.display = "none";
+                    });
+                    editor.addEventListener(Wodo.EVENT_HASLOCALUNSYNCEDOPERATIONSCHANGED, function(has) {
+                        hasLocalUnsyncedOpsOverlay.style.display = has ? "" : "none";
+                    });
+                    editor.addEventListener(Wodo.EVENT_HASSESSIONHOSTCONNECTIONCHANGED, function(has) {
+                        disconnectedOverlay.style.display = has ? "none" : "";
+                    });
+                    editor.addEventListener(Wodo.EVENT_UNKNOWNERROR, handleEditingError);
+
+                    // load the document and get called back when it's live
+                    editor.joinSession(serverFactory.createSessionBackend(sessionId, memberId, server), onEditingStarted);
+                });
+            } else {
+                editor.joinSession(serverFactory.createSessionBackend(sessionId, memberId, server), onEditingStarted);
+            }
+        }, function(err) {
+            // TODO: handle error
+            runtime.log(err);
+        });
+    }
+
+    function loginSuccess(userData) {
+        runtime.log("connected:" + userData.full_name);
+        token = userData.securityToken || null;
+
+        if (args.sessionId) {
+            joinSession(args.sessionId);
+        }
+    }
+
+    window.jsGlobalInstance.registerUserData(args.username, args.userData);
+
+    // start the editor with network
+    runtime.log("starting collaborative editor for [jsglobal].");
+    require(["webodf/editor/backend/jsglobal/ServerFactory"], function (ServerFactory) {
+        serverFactory = new ServerFactory();
+        server = serverFactory.createServer();
+        // wait for a network connection to establish.
+        // call the callback when done, when finally failed, or
+        // when a timeout reached.
+        // the parameter to the callback is a string with the possible
+        // values: "unavailable", "timeout", "ready"
+        server.connect(8000, function (state) {
+            if (state === "ready") {
+                server.login(args.username, loginSuccess);
+            }
+        });
+    });
+}
+
+window.JsGlobalServer = (function() {
```

Comment:
Hrm :unamused: 

My experience is that once a copy&paste is introduced, it takes many times more effort to remove.